### ISSUE TITLE
Diversity plugin connection to secondary replica set of MongoDB fails in version 1.12

### ISSUE THREAD

**Issue by zzakir** [vernemq/vernemq#1821]
### Environment

- VerneMQ Version: 1.12
- OS: Ubuntu 20.04
- Erlang/OTP version (if building from source): N/A
- VerneMQ configuration (vernemq.conf) or the changes from the default
- Cluster size/standalone: Standalone

### Expected behaviour

In version 1.12 rc1 I was able to successfully connect to secondary or primary instance of a MongoDB replica set to authenticate a client.

### Actual behaviour

This is no longer the case in version 1.12. Here I am only able to connect to the primary instance of the MongoDB replica set. Seems something was removed from version 1.12 rc1 before the release of final version 1.12.

**Comment by ioolkos** [comment]
@zzakir hm, ok, there was this fixes: https://github.com/vernemq/vernemq/commit/178a7afbc88d8d3698a4ff9f725bb9d881f87cc3
@varnerac does this ring a bell to you?

**Comment by varnerac** [comment]
@zzakir did you set `r_mode` to `slave_ok`?

any errors in the logs?

**Comment by zzakir** [comment]
@varnerac Yes, I did set r_mode to slave_ok in mongodb.lua file. This is the error which I get.

`2021-06-14 10:26:15.810 [error] <0.424.0> Supervisor {<0.424.0>,poolboy_sup} had child vmq_diversity_worker_wrapper started with {vmq_diversity_worker_wrapper,start_link,undefined} at <0.1012.0> exit with reason not_master in mc_connection_man:process_error/2 line 119 in context child_terminated`

When I do it with 1.12 rc 1, it works properly.

**Comment by varnerac** [comment]
Thanks. Let me walk the changes between those 1.12 rc 1 and 1.12, including the upstream Erlang MongoDB driver, and see if I can figure this out.

**Comment by varnerac** [comment]
out of curiosity, do you have `r_mode` set in the VMQ config? Could you try setting it there too?

**Comment by zzakir** [comment]
We can set r_mode via VMQ config as well? I was not aware of that. Where do we set this? I will try it and get back to you tomorrow.

**Comment by varnerac** [comment]
Here's the description of it in the VMQ Diversity config schema:

https://github.com/vernemq/vernemq/blob/master/apps/vmq_diversity/priv/vmq_diversity.schema#L297-L300

It's hidden =-)

**Comment by varnerac** [comment]
Also, it may not solve your issue. I just want to rule things out.

**Comment by zzakir** [comment]
Setting r_mode to slave_ok in the VMQ config does not solve the issue. So I think there is a bug here.

**Comment by varnerac** [comment]
Thanks for trying that.

**Comment by varnerac** [comment]
I'm gonna try to tackle the issue this week. I'll give any updates I find.

**Comment by varnerac** [comment]
I believe [this](https://github.com/vernemq/vernemq/pull/1848) will fix it. Waiting on tests to run.

**Comment by zzakir** [comment]
Thanks for the update @varnerac. I will test it once the new version with the fix is released.

**Comment by varnerac** [comment]
So, this is gonna be more of a fix that I thought. I'm going in and adding an automated test suite that includes a replica set scenario to the upstream MongoDB driver. I suspect it'll take 2 weeks or so.

**Comment by varnerac** [comment]
[This](https://github.com/comtihon/mongodb-erlang/issues/244) seems to be the issue. I'm looking at the best path forward after getting feedback from the driver maintainer. I bet the answer will be to pass a read preference. I need to learn more about nested tables in Lua to see if it's possible to pass `readopts` to the driver from the VMQ Diversity script.

**Comment by varnerac** [comment]
This should fix it: https://github.com/vernemq/vernemq/pull/1877

**Comment by zzakir** [comment]
@varnerac This fix is part of release 1.12.3 right? If so, I will test it today

**Comment by ioolkos** [comment]
@zzakir yes, that's merged into 1.12.3. Thanks for testing!

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by zzakir** [comment]
@varnerac I just tested it. Vernemq is able to connect to both primary and secondary instances of MongoDB without any hiccups. And I didn't even have to update the lua file like I had to the last time. It just works! Great work and thanks a lot!

I guess you can close this issue now right?
